From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 2 May 2026 08:31:52 -0400
Subject: [PATCH] fixup: Leaves: Lithium Sleeping Block Entity

This is a temp patch for tracking new changes added in Lithium that Leaves haven't updated yet.
This patch will merge into Leaves once they finish the 26.x update, and sync back to Leaf,
(merge into original patch).

Last checked Lithium commit: f081ee429879b7c395c22843a41d5ee5b4379ede

Commits/changes synced from Lithium:
https://github.com/CaffeineMC/lithium/commit/d2f48f22cb3890e81bcb49ba61dc625685b28c86
https://github.com/CaffeineMC/lithium/commit/9d29463e79e821ab3d80ade274312caab0a6d397
https://github.com/CaffeineMC/lithium/commit/2c2da310d5e21fe23a8886030503c68d8adfae4e
https://github.com/CaffeineMC/lithium/commit/d2f48f22cb3890e81bcb49ba61dc625685b28c86
https://github.com/CaffeineMC/lithium/commit/d2f48f22cb3890e81bcb49ba61dc625685b28c86
https://github.com/CaffeineMC/lithium/commit/26985fca69279a3bb735d8f0b21f7569906b0870
https://github.com/CaffeineMC/lithium/commit/f1a897deff1ac25884ea66923f9c8dbdbd25ee54
https://github.com/CaffeineMC/lithium/commit/66183a05f63362c40633bae3161d27c5a8af3ba6
https://github.com/CaffeineMC/lithium/commit/59b8dcb226a78c04964aec69cbb3d745ed0515a7
https://github.com/CaffeineMC/lithium/commit/990618bb6d1faebf6fc5f8c43105d299af857d13

diff --git a/net/minecraft/world/level/block/SculkSensorBlock.java b/net/minecraft/world/level/block/SculkSensorBlock.java
index 4737afb0f5b149f0286639353ab2a42d6e181c59..ec7b080cf3d3f77e16194e5eef139de28ec6570f 100644
--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -168,7 +168,17 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
         return createTickerHelper(
                 blockEntityType,
                 BlockEntityType.SCULK_SENSOR,
-                (level1, pos, state1, blockEntity) -> VibrationSystem.Ticker.tick(level1, blockEntity.getVibrationData(), blockEntity.getVibrationUser())
+                // Leaves start - Lithium Sleeping Block Entity
+                (level1, pos, state1, blockEntity) -> {
+                    final VibrationSystem.Data vibrationData = blockEntity.getVibrationData();
+                    VibrationSystem.Ticker.tick(level1, vibrationData, blockEntity.getVibrationUser());
+                    if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled
+                        && vibrationData.getCurrentVibration() == null
+                        && vibrationData.getSelectionStrategy().chosenCandidate(Long.MAX_VALUE).isEmpty()) {
+                        blockEntity.lithium$startSleeping();
+                    }
+                }
+                // Leaves end - Lithium Sleeping Block Entity
             );
     }
 
diff --git a/net/minecraft/world/level/block/SculkShriekerBlock.java b/net/minecraft/world/level/block/SculkShriekerBlock.java
index 857a5895a3d95014f661f041ef54f259750ecca1..f5ce1fbe6b93e7a50e1a5b6a990625370f82b137 100644
--- a/net/minecraft/world/level/block/SculkShriekerBlock.java
+++ b/net/minecraft/world/level/block/SculkShriekerBlock.java
@@ -144,7 +144,17 @@ public class SculkShriekerBlock extends BaseEntityBlock implements SimpleWaterlo
         return BaseEntityBlock.createTickerHelper(
                 blockEntityType,
                 BlockEntityType.SCULK_SHRIEKER,
-                (level1, pos, state1, blockEntity) -> VibrationSystem.Ticker.tick(level1, blockEntity.getVibrationData(), blockEntity.getVibrationUser())
+                // Leaves start - Lithium Sleeping Block Entity
+                (level1, pos, state1, blockEntity) -> {
+                    final VibrationSystem.Data vibrationData = blockEntity.getVibrationData();
+                    VibrationSystem.Ticker.tick(level1, vibrationData, blockEntity.getVibrationUser());
+                    if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled
+                        && vibrationData.getCurrentVibration() == null
+                        && vibrationData.getSelectionStrategy().chosenCandidate(Long.MAX_VALUE).isEmpty()) {
+                        blockEntity.lithium$startSleeping();
+                    }
+                }
+                // Leaves end - Lithium Sleeping Block Entity
             );
     }
 }
diff --git a/net/minecraft/world/level/block/SculkSpreader.java b/net/minecraft/world/level/block/SculkSpreader.java
index 3485a08070b1f021ee0dfb9b7456d0227599027b..3ba034ebe82ba078a2eb000aeb583663fc8384f1 100644
--- a/net/minecraft/world/level/block/SculkSpreader.java
+++ b/net/minecraft/world/level/block/SculkSpreader.java
@@ -33,7 +33,7 @@ import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import org.jspecify.annotations.Nullable;
 
-public class SculkSpreader {
+public class SculkSpreader implements net.caffeinemc.mods.lithium.common.block.entity.sleeping_sculk.GameEventListenerWithCallback { // Leaves - Lithium Sleeping Block Entity
     public static final int MAX_GROWTH_RATE_RADIUS = 24;
     public static final int MAX_CHARGE = 1000;
     public static final float MAX_DECAY_FACTOR = 0.5F;
@@ -141,6 +141,11 @@ public class SculkSpreader {
             }
             // CraftBukkit end
             this.cursors.add(cursor);
+            // Leaves start - Lithium Sleeping Block Entity
+            if (this.listener != null) {
+                this.listener.run();
+            }
+            // Leaves end - Lithium Sleeping Block Entity
         }
     }
 
@@ -190,6 +195,15 @@ public class SculkSpreader {
         }
     }
 
+    // Leaves start - Lithium Sleeping Block Entity
+    private Runnable listener;
+
+    @Override
+    public void lithium$setGameEventCallback(Runnable listener) {
+        this.listener = listener;
+    }
+    // Leaves end - Lithium Sleeping Block Entity
+
     public static class ChargeCursor {
         private static final ObjectArrayList<Vec3i> NON_CORNER_NEIGHBOURS = Util.make(
             new ObjectArrayList<>(18),
diff --git a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
index a613291f867a1c69421cf72ddb0a3d0c7c3534a3..ce6952bed7320fa52c673e5b8222a9109585efec 100644
--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -207,12 +207,16 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
     @Override
     public void lithium$emitStackListReplaced() {
         it.unimi.dsi.fastutil.objects.ReferenceArraySet<net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener> listeners = this.inventoryHandlingTypeListeners;
+        this.inventoryHandlingTypeListeners = null; //Prevent concurrent modification
         if (listeners != null && !listeners.isEmpty()) {
             for (net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener inventoryChangeListener : listeners) {
                 inventoryChangeListener.handleStackListReplaced(this);
             }
             listeners.clear();
         }
+        if (this.inventoryHandlingTypeListeners == null) {
+            this.inventoryHandlingTypeListeners = listeners;
+        }
 
         if (this instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener listener) {
             listener.handleStackListReplaced(this);
@@ -224,12 +228,16 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
     @Override
     public void lithium$emitRemoved() {
         it.unimi.dsi.fastutil.objects.ReferenceArraySet<net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener> listeners = this.inventoryHandlingTypeListeners;
+        this.inventoryHandlingTypeListeners = null; //Prevent concurrent modification
         if (listeners != null && !listeners.isEmpty()) {
             for (net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener listener : listeners) {
                 listener.lithium$handleInventoryRemoved(this);
             }
             listeners.clear();
         }
+        if (this.inventoryHandlingTypeListeners == null) {
+            this.inventoryHandlingTypeListeners = listeners;
+        }
 
         if (this instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener listener) {
             listener.lithium$handleInventoryRemoved(this);
@@ -258,11 +266,11 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
     }
 
     @Override
-    public void lithium$forwardContentChangeOnce(net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener inventoryChangeListener, net.caffeinemc.mods.lithium.common.hopper.LithiumStackList stackList, net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker thisTracker) {
+    public void lithium$forwardContentChangeOnce(net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener inventoryChangeListener, net.caffeinemc.mods.lithium.common.hopper.LithiumStackList stackList) {
         if (this.inventoryChangeListeners == null) {
             this.inventoryChangeListeners = new it.unimi.dsi.fastutil.objects.ReferenceArraySet<>(1);
         }
-        stackList.setInventoryModificationCallback(thisTracker);
+        stackList.setNextInventoryModificationCallback((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this);
         this.inventoryChangeListeners.add(inventoryChangeListener);
 
     }
diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index f0db30e77584eed00abdf3067326600f5c6bca97..9d4931ca0ea1b929c66bdbf42c0b33fe77d83af2 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -972,10 +972,17 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 Container blockInventory = this.extractBlockInventory;
                 if (this.extractStackList != null &&
                     blockInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) {
-                    if (!this.extractStackList.maybeSendsComparatorUpdatesOnFailedExtract() || (blockInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_comparator_tracking.ComparatorTracker comparatorTracker && !comparatorTracker.lithium$hasAnyComparatorNearby())) {
-                        listenToExtractTracker = true;
+                    if (this.extractStackList.maybeSendsComparatorUpdatesOnFailedExtract() && this.extractStackList.getOccupiedSlots() != 0) {
+                        if (blockInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_comparator_tracking.ComparatorTracker comparatorTracker && !comparatorTracker.lithium$hasAnyComparatorNearby()) {
+                            listenToExtractTracker = true;
+                        } else {
+                            //Inventory is not empty (0 != number of occupied slots) and maybe sends comparator
+                            // updates on failed extract attempts, so hopper must not sleep to be able to send
+                            // the observable comparator updates.
+                            return;
+                        }
                     } else {
-                        return;
+                        listenToExtractTracker = true;
                     }
                 } else {
                     return;
@@ -1154,16 +1161,6 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         }
     }
 
-    private void invalidateBlockInsertionData() {
-        this.insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
-        this.insertBlockInventory = null;
-        this.insertInventory = null;
-        this.insertStackList = null;
-        this.insertStackListModCount = 0;
-
-        wakeUpNow();
-    }
-
     private void invalidateCachedData() {
         this.shouldCheckSleep = false;
         this.invalidateInsertionData();
@@ -1180,11 +1177,22 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             }
         }
 
+        this.invalidateBlockInsertionData();
+    }
+
+    private void invalidateBlockInsertionData() {
         if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
             assert this.insertBlockInventory != null;
             ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.insertBlockInventory).stopListenForMajorInventoryChanges(this);
         }
-        this.invalidateBlockInsertionData();
+
+        this.insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+        this.insertBlockInventory = null;
+        this.insertInventory = null;
+        this.insertStackList = null;
+        this.insertStackListModCount = 0;
+
+        wakeUpNow();
     }
 
     private void invalidateExtractionData() {
@@ -1202,14 +1210,15 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 this.collectItemEntityTrackerWasEmpty = false;
             }
         }
+        this.invalidateBlockExtractionData();
+    }
+
+    private void invalidateBlockExtractionData() {
         if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
             assert this.extractBlockInventory != null;
             ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.extractBlockInventory).stopListenForMajorInventoryChanges(this);
         }
-        this.invalidateBlockExtractionData();
-    }
 
-    private void invalidateBlockExtractionData() {
         this.extractionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
         this.extractBlockInventory = null;
         this.extractInventory = null;
diff --git a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index 79e5a0ed25ec40391a0b78f309eabfc9fa9f0de9..70e84a9eb442d93c0d7e6ab5e8e0c788b168ed36 100644
--- a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -27,12 +27,13 @@ import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
 
-public class SculkCatalystBlockEntity extends BlockEntity implements GameEventListener.Provider<SculkCatalystBlockEntity.CatalystListener> {
+public class SculkCatalystBlockEntity extends BlockEntity implements GameEventListener.Provider<SculkCatalystBlockEntity.CatalystListener>, net.caffeinemc.mods.lithium.common.block.entity.SleepingBlockEntity { // Leaves - Lithium Sleeping Block Entity
     private final SculkCatalystBlockEntity.CatalystListener catalystListener;
 
     public SculkCatalystBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.SCULK_CATALYST, pos, blockState);
         this.catalystListener = new SculkCatalystBlockEntity.CatalystListener(blockState, new BlockPositionSource(pos));
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled) this.catalystListener.getSculkSpreader().lithium$setGameEventCallback(this::wakeUpNow); // Leaves - Lithium Sleeping Block Entity
     }
 
     // Paper start - Fix NPE in SculkBloomEvent world access
@@ -47,12 +48,22 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
         org.bukkit.craftbukkit.event.CraftEventFactory.setSourceBlockOverrideRT(pos); // CraftBukkit - SPIGOT-7068: Add source block override, not the most elegant way but better than passing down a BlockPosition up to five methods deep. // Leaf - SparklyPaper - parallel world ticking
         sculkCatalyst.catalystListener.getSculkSpreader().updateCursors(level, pos, level.getRandom(), true);
         org.bukkit.craftbukkit.event.CraftEventFactory.setSourceBlockOverrideRT(null); // CraftBukkit //  Leaf - SparklyPaper - parallel world ticking
+        // Leaves start - Lithium Sleeping Block Entity
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled && sculkCatalyst.getListener().getSculkSpreader().getCursors().isEmpty()) {
+            sculkCatalyst.lithium$startSleeping();
+        }
+        // Leaves end - Lithium Sleeping Block Entity
     }
 
     @Override
     protected void loadAdditional(ValueInput input) {
         super.loadAdditional(input);
         this.catalystListener.sculkSpreader.load(input);
+        // Leaves start - Lithium Sleeping Block Entity
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled && !this.getListener().getSculkSpreader().getCursors().isEmpty()) {
+            this.wakeUpNow();
+        }
+        // Leaves end - Lithium Sleeping Block Entity
     }
 
     @Override
@@ -66,6 +77,32 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
         return this.catalystListener;
     }
 
+    // Leaves start - Lithium Sleeping Block Entity
+    private net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public TickingBlockEntity lithium$getSleepingTicker() {
+        return sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Leaves end - Lithium Sleeping Block Entity
+
     public class CatalystListener implements GameEventListener { // Leaf - petal - reduce work done by game event system
         public static final int PULSE_TICKS = 8;
         final SculkSpreader sculkSpreader;
diff --git a/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java b/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
index 39aa9e1703ade12899f26ffc8063895e52138adb..21daebdccc74750ece1ffb6b94bc5d6e9c363f99 100644
--- a/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkSensorBlockEntity.java
@@ -15,7 +15,7 @@ import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import org.jspecify.annotations.Nullable;
 
-public class SculkSensorBlockEntity extends BlockEntity implements GameEventListener.Provider<VibrationSystem.Listener>, VibrationSystem {
+public class SculkSensorBlockEntity extends BlockEntity implements GameEventListener.Provider<VibrationSystem.Listener>, VibrationSystem, net.caffeinemc.mods.lithium.common.block.entity.SleepingBlockEntity { // Leaves - Lithium Sleeping Block Entity
     private static final int DEFAULT_LAST_VIBRATION_FREQUENCY = 0;
     private VibrationSystem.Data vibrationData;
     private final VibrationSystem.Listener vibrationListener;
@@ -28,6 +28,7 @@ public class SculkSensorBlockEntity extends BlockEntity implements GameEventList
         this.vibrationUser = this.createVibrationUser();
         this.vibrationData = new VibrationSystem.Data();
         this.vibrationListener = new VibrationSystem.Listener(this);
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled) this.vibrationListener.lithium$setGameEventCallback(this::wakeUpNow); // Leaves - Lithium Sleeping Block Entity
     }
 
     public SculkSensorBlockEntity(BlockPos pos, BlockState blockState) {
@@ -44,6 +45,11 @@ public class SculkSensorBlockEntity extends BlockEntity implements GameEventList
         this.lastVibrationFrequency = input.getIntOr("last_vibration_frequency", 0);
         this.vibrationData = input.read("listener", VibrationSystem.Data.CODEC).orElseGet(VibrationSystem.Data::new);
         this.rangeOverride = input.getInt(PAPER_LISTENER_RANGE_NBT_KEY).orElse(null); // Paper start - Configurable sculk sensor listener range
+        // Leaves start - Lithium Sleeping Block Entity
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled && vibrationData.getSelectionStrategy().chosenCandidate(Long.MAX_VALUE).isPresent()) {
+            this.wakeUpNow();
+        }
+        // Leaves end - Lithium Sleeping Block Entity
     }
 
     protected static final String PAPER_LISTENER_RANGE_NBT_KEY = "Paper.ListenerRange"; // Paper - Configurable sculk sensor listener range
@@ -83,6 +89,32 @@ public class SculkSensorBlockEntity extends BlockEntity implements GameEventList
         return this.vibrationListener;
     }
 
+    // Leaves start - Lithium Sleeping Block Entity
+    private net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public TickingBlockEntity lithium$getSleepingTicker() {
+        return sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Leaves end - Lithium Sleeping Block Entity
+
     protected class VibrationUser implements VibrationSystem.User {
         public static final int LISTENER_RANGE = 8;
         protected final BlockPos blockPos;
diff --git a/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java b/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
index 8a999d8e72b306d1e4a371c7a43cdf6e5b13daff..30e711cb4272cda6f0ef159f2f40080adab54b10 100644
--- a/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -39,7 +39,7 @@ import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public class SculkShriekerBlockEntity extends BlockEntity implements GameEventListener.Provider<VibrationSystem.Listener>, VibrationSystem {
+public class SculkShriekerBlockEntity extends BlockEntity implements GameEventListener.Provider<VibrationSystem.Listener>, VibrationSystem, net.caffeinemc.mods.lithium.common.block.entity.SleepingBlockEntity { // Leaves - Lithium Sleeping Block Entity
     private static final int WARNING_SOUND_RADIUS = 10;
     private static final int WARDEN_SPAWN_ATTEMPTS = 20;
     private static final int WARDEN_SPAWN_RANGE_XZ = 5;
@@ -60,6 +60,7 @@ public class SculkShriekerBlockEntity extends BlockEntity implements GameEventLi
 
     public SculkShriekerBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.SCULK_SHRIEKER, pos, blockState);
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled) this.vibrationListener.lithium$setGameEventCallback(this::wakeUpNow); // Leaves - Lithium Sleeping Block Entity
     }
 
     @Override
@@ -77,6 +78,11 @@ public class SculkShriekerBlockEntity extends BlockEntity implements GameEventLi
         super.loadAdditional(input);
         this.warningLevel = input.getIntOr("warning_level", 0);
         this.vibrationData = input.read("listener", VibrationSystem.Data.CODEC).orElseGet(VibrationSystem.Data::new);
+        // Leaves start - Lithium Sleeping Block Entity
+        if (org.dreeam.leaf.config.modules.opt.SleepingBlockEntity.enabled && this.vibrationData.getSelectionStrategy().chosenCandidate(Long.MAX_VALUE).isPresent()) {
+            this.wakeUpNow();
+        }
+        // Leaves end - Lithium Sleeping Block Entity
     }
 
     @Override
@@ -179,6 +185,32 @@ public class SculkShriekerBlockEntity extends BlockEntity implements GameEventLi
         return this.vibrationListener;
     }
 
+    // Leaves start - Lithium Sleeping Block Entity
+    private net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public TickingBlockEntity lithium$getSleepingTicker() {
+        return sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Leaves end - Lithium Sleeping Block Entity
+
     class VibrationUser implements VibrationSystem.User {
         private static final int LISTENER_RADIUS = 8;
         private final PositionSource positionSource = new BlockPositionSource(SculkShriekerBlockEntity.this.worldPosition);
diff --git a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
index 3f5ffa8679c6ba6684175938d41d5a947063ebf6..8ee5ec279acde39c6137dc72bafa1aa1d6184fd3 100644
--- a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
+++ b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
@@ -181,7 +181,7 @@ public interface VibrationSystem {
         }
     }
 
-    public static class Listener implements GameEventListener {
+    public static class Listener implements GameEventListener, net.caffeinemc.mods.lithium.common.block.entity.sleeping_sculk.GameEventListenerWithCallback { // Leaves - Lithium Sleeping Block Entity
         private final VibrationSystem system;
 
         public Listener(VibrationSystem system) {
@@ -225,6 +225,11 @@ public interface VibrationSystem {
                         return false;
                     } else {
                         this.scheduleVibration(level, vibrationData, gameEvent, context, pos, vec3);
+                        // Leaves start - Lithium Sleeping Block Entity
+                        if (this.listener != null) {
+                            this.listener.run();
+                        }
+                        // Leaves end - Lithium Sleeping Block Entity
                         return true;
                     }
                 }
@@ -264,6 +269,15 @@ public interface VibrationSystem {
 
             return true;
         }
+
+        // Leaves start - Lithium Sleeping Block Entity
+        private Runnable listener;
+
+        @Override
+        public void lithium$setGameEventCallback(Runnable listener) {
+            this.listener = listener;
+        }
+        // Leaves end - Lithium Sleeping Block Entity
     }
 
     public interface Ticker {
